glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,543 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the MIT/Apache-2.0 License, at your convenience
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+//
+
+use crate::executor::TaskQueueHandle;
+use nix::sys;
+use std::{
+    fmt,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::{self, JoinHandle},
+    time::{Duration, Instant},
+};
+
+pub struct StallDetection<'a> {
+    executor: usize,
+    queue_handle: TaskQueueHandle,
+    queue_name: &'a str,
+    trace: backtrace::Backtrace,
+    budget: Duration,
+    overage: Duration,
+}
+
+impl fmt::Debug for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("StallDetection")
+            .field("executor", &self.executor)
+            .field("queue_handle", &self.queue_handle)
+            .field("queue_name", &self.queue_name)
+            .field("trace", &self.trace)
+            .field("budget", &self.budget)
+            .field("overage", &self.overage)
+            .finish()
+    }
+}
+
+impl fmt::Display for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[stall-detector -- executor {}] task queue {} went over-budget: {:#?} (budget: \
+             {:#?}). Backtrace: {:#?}",
+            self.executor, self.queue_name, self.overage, self.budget, self.trace,
+        )
+    }
+}
+
+/// Trait describing what signal to use to trigger stall detection,
+/// how far past expected execution time to trigger a stall,
+/// and how to handle a stall detection once triggered.
+pub trait StallDetectionHandler: std::fmt::Debug + Send + Sync {
+    /// How far past the preemption timer should qualify as a stall
+    /// If None is returned, don't use the stall detector for this task queue.
+    fn high_water_mark(
+        &self,
+        queue_handle: TaskQueueHandle,
+        max_expected_runtime: Duration,
+    ) -> Option<Duration>;
+
+    /// What signal number to use; see values in libc::SIG*
+    fn signal(&self) -> u8;
+
+    /// Handler called when a task exceeds its budget
+    fn stall(&self, detection: StallDetection<'_>);
```

Comment:
Let's add a comment to explain this callback is invoked from inside the reactor. We should ask users to not perform blocking syscalls, not to acquire locks, not to spawn tasks, etc.

Reply:
In fact, I would be almost tempted to unset the LOCAL_EX TLS var just so that trying to enter the executor would panic the app.